rules/ast/ast.go

```diff
@@ -749,55 +756,185 @@ func labelsEqual(labels1, labels2 clientmodel.Metric) bool {
 // Eval implements the VectorNode interface and returns the result of
 // the expression.
 func (node *VectorArithExpr) Eval(timestamp clientmodel.Timestamp) Vector {
-	result := Vector{}
-	if node.lhs.Type() == ScalarType && node.rhs.Type() == VectorType {
-		lhs := node.lhs.(ScalarNode).Eval(timestamp)
+	// vector-to-vector operation
+	if node.lhs.Type() == VectorType && node.rhs.Type() == VectorType {
+		lhs := node.lhs.(VectorNode).Eval(timestamp)
 		rhs := node.rhs.(VectorNode).Eval(timestamp)
-		for _, rhsSample := range rhs {
-			value, keep := evalVectorBinop(node.opType, lhs, rhsSample.Value)
-			if keep {
-				rhsSample.Value = value
-				if node.opType.shouldDropMetric() {
-					rhsSample.Metric.Delete(clientmodel.MetricNameLabel)
-				}
-				result = append(result, rhsSample)
+
+		return node.evalVectors(timestamp, lhs, rhs)
+	}
+
+	// vector-to-scalar operation
+	var lhs Vector
+	var rhs clientmodel.SampleValue
+	swap := false
+
+	if node.lhs.Type() == ScalarType && node.rhs.Type() == VectorType {
+		lhs = node.rhs.(VectorNode).Eval(timestamp)
+		rhs = node.lhs.(ScalarNode).Eval(timestamp)
+		swap = true
+	} else {
+		lhs = node.lhs.(VectorNode).Eval(timestamp)
+		rhs = node.rhs.(ScalarNode).Eval(timestamp)
+	}
+
+	result := make(Vector, 0, len(lhs))
+
+	for _, lhsSample := range lhs {
+		lv, rv := lhsSample.Value, rhs
+		// lhs always contains the vector. If the original position was different
+		// swap for calculating the value.
+		if swap {
+			lv, rv = rv, lv
+		}
+		value, keep := evalVectorBinop(node.opType, lv, rv)
+		if keep {
+			lhsSample.Value = value
+			if node.opType.shouldDropMetric() {
+				lhsSample.Metric.Delete(clientmodel.MetricNameLabel)
 			}
+			result = append(result, lhsSample)
 		}
-		return result
-	} else if node.lhs.Type() == VectorType && node.rhs.Type() == ScalarType {
-		lhs := node.lhs.(VectorNode).Eval(timestamp)
-		rhs := node.rhs.(ScalarNode).Eval(timestamp)
-		for _, lhsSample := range lhs {
-			value, keep := evalVectorBinop(node.opType, lhsSample.Value, rhs)
-			if keep {
-				lhsSample.Value = value
-				if node.opType.shouldDropMetric() {
-					lhsSample.Metric.Delete(clientmodel.MetricNameLabel)
-				}
-				result = append(result, lhsSample)
+	}
+	return result
+}
+
+// evalVectorBinop evaluates the binary operation for the given vectors.
```

Comment:
Hmm, maybe this should actually be called `evalVectorBinop` now, and the other function should be renamed?